File: receptor/link/link.go
// Capa de Enlace - Orquestador | Receptor
// Responsabilidad: parsear la trama recibida, identificar el algoritmo
// y delegar la verificación/corrección al módulo correspondiente.

package link

import (
	"fmt"
	"strings"
)

// ResultadoEnlace contiene el resultado del procesamiento de la capa de enlace.
type ResultadoEnlace struct {
	BitsMensaje string
	HayError    bool
	MensajeError string
}

// VerificarIntegridad parsea la trama, detecta el algoritmo y verifica/corrige el frame.
// Retorna un ResultadoEnlace con los bits del mensaje o la descripción del error.
func VerificarIntegridad(trama string) ResultadoEnlace {
	fmt.Printf("\n[ENLACE] Procesando trama recibida...\n")

	// Parsear el formato "algoritmo|bits"
	partes := strings.SplitN(trama, "|", 2)
	if len(partes) != 2 {
		return ResultadoEnlace{
			HayError:     true,
			MensajeError: fmt.Sprintf("formato de trama inválido: se esperaba 'algoritmo|bits', se recibió: %q", trama),
		}
	}

	algoritmo := strings.TrimSpace(partes[0])
	frame := strings.TrimSpace(partes[1])

	fmt.Printf("[ENLACE] Algoritmo detectado: %s\n", strings.ToUpper(algoritmo))
	fmt.Printf("[ENLACE] Frame a procesar: %d bits\n", len(frame))

	switch algoritmo {
	case "hamming":
		bits, err := DecodificarHamming(frame)
		if err != nil {
			return ResultadoEnlace{
				HayError:     true,
				MensajeError: fmt.Sprintf("Hamming: %v", err),
			}
		}
		return ResultadoEnlace{BitsMensaje: bits, HayError: false}

	case "crc32":
		bits, err := VerificarCRC32(frame)
		if err != nil {
			return ResultadoEnlace{
				HayError:     true,
				MensajeError: fmt.Sprintf("CRC-32: %v", err),
			}
		}
		return ResultadoEnlace{BitsMensaje: bits, HayError: false}

	default:
		return ResultadoEnlace{
			HayError:     true,
			MensajeError: fmt.Sprintf("algoritmo desconocido: %q. Se esperaba 'hamming' o 'crc32'", algoritmo),
		}
	}
}

File: emisor/link/link.py
# Capa de Enlace - Orquestador | Emisor
# Responsabilidad: según el algoritmo elegido, calcular la integridad
# y retornar el frame listo para enviarse (con el algoritmo indicado como prefijo).


from link.hamming import codificar_hamming
from link.crc32 import calcular_crc32

# Recibe el mensaje en bits y el algoritmo a usar.
# Retorna el frame completo con prefijo de algoritmo.
def calcular_integridad(bits_mensaje: str, algoritmo: str) -> str:
    print(f"\n[ENLACE] Calculando integridad con algoritmo: {algoritmo.upper()}")

    if algoritmo == "hamming":
        frame = codificar_hamming(bits_mensaje)

    elif algoritmo == "crc32":
        frame = calcular_crc32(bits_mensaje)

    else:
        raise ValueError(f"Algoritmo desconocido: '{algoritmo}'. Use 'hamming' o 'crc32'.")

    trama_final = f"{algoritmo}|{frame}"
    print(f"[ENLACE] Trama lista para transmisión: {algoritmo}|[{len(frame)} bits]")
    return trama_final